modules/core/react/lib/InputProvider.tsx

```diff
@@ -150,67 +146,76 @@ export default class InputProvider extends React.Component<InputProviderProps, I
       console.warn('Browser does not support passive event listeners');
     }
 
-    this.state = {
-      currentInput: storedInput || InputType.Initial,
-      currentIntent: storedIntent || InputType.Initial,
-      supportsPassive: supportsPassive || false,
-      isBuffering: false,
-      isScrolling: false,
-      mousePosX: null,
-      mousePosY: null,
-      nested: false,
-    };
+    this.currentInput = storedInput || InputType.Initial;
+    this.currentIntent = storedIntent || InputType.Initial;
+    this.supportsPassive = supportsPassive || false;
 
     this.setInput = this.setInput.bind(this);
     this.setIntent = this.setIntent.bind(this);
     this.eventBuffer = this.eventBuffer.bind(this);
   }
 
+  static defaultProps = {
+    container: document.body,
+  };
+
   // Need to remember how it component was mounted so we ensure listener removal
   provideIntent = this.props.provideIntent;
 
+  getContainer(container: HTMLElement | React.RefObject<HTMLElement>): HTMLElement {
+    if ('current' in container) {
+      if (container.current === null) {
+        console.warn('Your ref object can not be null, therefore, falling back to document.body');
+        return document.body;
+      } else {
+        return container.current;
+      }
+    }
+    return container;
+  }
+
+  private container = this.getContainer(this.props.container);
+
   componentDidMount() {
     // For IE11 and under, we'll need to polyfill element.closest
     elementClosestPolyfill(window);
 
-    if (
-      this.ref.current &&
-      this.ref.current.parentElement &&
-      this.ref.current.parentElement.closest(`.${containerClass}`)
-    ) {
-      this.setState({nested: true});
+    if (this.container.closest('[data-whatinput]')) {
+      this.isAnotherInputProvider = true;
       return;
     }
+    this.container.setAttribute('data-whatinput', this.currentInput);
+    this.container.setAttribute('data-whatintent', this.currentIntent);
```

Comment:
We should only add intent if `this.provideIntent`. Might as well just call `updateAttributes()` here
```suggestion
    this.updateAttributes()
```